resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/stream/config/AkkaRunnableGraph.java

```diff
@@ -0,0 +1,127 @@
+import akka.stream.ActorAttributes;
+import akka.stream.Materializer;
+import akka.stream.Supervision;
+import akka.stream.Supervision.Directive;
+import akka.stream.javadsl.RunnableGraph;
+import akka.stream.javadsl.Sink;
+import akka.stream.javadsl.Source;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import scala.Function1;
+
+import uk.gov.gchq.palisade.service.resource.model.ResourceRequest;
+import uk.gov.gchq.palisade.service.resource.model.ResourceResponse;
+import uk.gov.gchq.palisade.service.resource.model.ResourceResponse.Builder;
+import uk.gov.gchq.palisade.service.resource.service.KafkaProducerService;
+import uk.gov.gchq.palisade.service.resource.service.StreamingResourceServiceProxy;
+import uk.gov.gchq.palisade.service.resource.stream.ConsumerTopicConfiguration;
+import uk.gov.gchq.palisade.service.resource.stream.ProducerTopicConfiguration;
+import uk.gov.gchq.palisade.service.resource.stream.ProducerTopicConfiguration.Topic;
+
+import java.util.Optional;
+import java.util.concurrent.CompletionStage;
+
+/**
+ * Configuration for the Akka Runnable Graph used by the {@link uk.gov.gchq.palisade.service.resource.ResourceApplication}
+ * Configures the connection between Kafka, Akka and the service
+ */
+@Configuration
+public class AkkaRunnableGraph {
+
+    @Bean
+    KafkaProducerService kafkaProducerService(final Sink<ProducerRecord<String, ResourceRequest>, CompletionStage<Done>> sink,
+                                              final ConsumerTopicConfiguration upstreamConfig,
+                                              final Materializer materializer) {
+        return new KafkaProducerService(sink, upstreamConfig, materializer);
+    }
+
+    @Bean
+    Function1<Throwable, Directive> supervisor() {
+        return ex -> Supervision.resumingDecider().apply(ex);
+    }
+
+    @Bean
+    RunnableGraph<DrainingControl<Done>> runner(
+            final Source<CommittableMessage<String, ResourceRequest>, Control> source,
+            final Sink<Envelope<String, ResourceResponse, Committable>, CompletionStage<Done>> sink,
+            final Function1<Throwable, Directive> supervisionStrategy,
+            final ProducerTopicConfiguration topicConfiguration,
+            final StreamingResourceServiceProxy service) {
+        // Get output topic from config
+        Topic outputTopic = topicConfiguration.getTopics().get("output-topic");
+
+        // Read messages from the stream source
+        return source
+                .map(message -> new Pair<>(message, Optional.ofNullable(message.record().value())))
+                // Get a stream of resources from an iterator
+                .flatMapConcat(messageAndRequest -> messageAndRequest.second()
+                        .map(request -> Source.fromIterator(() -> service.getResourcesById(request.resourceId))
```

Comment:
Shouldnt these be async calls?

Reply:
This map is being done on the `Request` so it's not possible to do a `.mapAsync`. It's creating an Akka stream from the return value of the request so not sure if anything needs/can be done.

Reply:
I have added some extra comments to the runnableGraph so that it has a better explanation of what each step is doing, especially the step of adding an `Optional.empty()` to the stream.

Reply:
I can see a case for having an async proxy and doing a `.mapAsync(msgAndReq -> new Pair<>(msgAndReq, msgAndReq.second().map(req -> asyncSvc.getResourcesById(req.resourceId))))` then continuing with the `.flatMapConcat()`
The control flow is a little trickier, but there's justification for it...